---
title: "Applying GLM-PCA to Data"
author: "Will Townes, Kelly Street"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r}
require(glmpca)
```

Simulate some data. Thanks to Jake Yeung for providing the [initial code for the simulation](https://github.com/willtownes/scrna2019/issues/2).

```{r}
ngenes <- 10000
ncells <- 50

# simulate two batches with different depths
ncounts <- rpois(ncells*3, lambda = 1000*rep(1:2, each = ncells*3/2))
# generate profiles for 3 clusters
profiles <- replicate(3, exp(rnorm(n = ngenes)))
# generate cluster labels
cluster <- sample(rep(1:3, each = ncells))
# generate single-cell transcriptomes 
counts <- sapply(seq_along(cluster), function(i){
	rmultinom(1, ncounts[i], prob = profiles[,cluster[i]])
})
rownames(counts) <- paste("gene", seq(nrow(counts)), sep = "_")
colnames(counts) <- paste("cell", seq(ncol(counts)), sep = "_")
# clean up rows
counts <- counts[rowSums(counts) > 0, ]

# run GLM-PCA
Yout <- glmpca(Y = counts, L=2, fam = "poi", verbose = TRUE) # fam="mult" gives same "ring"

colvec <- c('red3','dodgerblue3','green3')[cluster]
plot(Yout$factors, asp = 1, col = colvec, pch = 20, xlab = "GLMPC1 Loadings", ylab = "GLMPC2 Loadings", cex = 5, cex.lab = 1.5, cex.main = 1.5, main = "GLMPCA on Observed Space")
```

analyze the same data with PCA, the batch effect produces incorrect clusters.

```{r}
pca <- prcomp(t(log1p(counts)))

plot(pca$x, asp = 1, col = colvec, pch = 20, xlab = "PC1 Loadings", ylab = "PC2 Loadings", cex = 5, cex.lab = 1.5, cex.main = 1.5, main = "PCA on Observed Space")
```